game-core/src/main/java/org/triplea/performance/PerfTimer.java

```diff
@@ -39,36 +41,54 @@ private long stopTimer() {
 
   @Override
   public void close() {
-    processResult(stopTimer(), this);
+    if (this.reportingFrequency > 0) {
+      processResult(stopTimer(), this);
+    }
   }
 
   @SuppressWarnings("unused")
   public static PerfTimer startTimer(final String title) {
-    return new PerfTimer(title);
+    return startTimer(title, 1);
   }
 
-  private static synchronized void processResult(final long stopNanos, final PerfTimer perfTimer) {
-    final long stopMicros = stopNanos / 1000;
-
-    final long milliFraction = (stopMicros % 1000) / 100;
-    final long millis = (stopMicros / 1000);
+  public static PerfTimer startTimer(final String title, final int reportingFrequency) {
```

Comment:
A javadoc here to describe that reporting frequency means we report every `n'th` could be useful to people down the road wanting to use this API.

Reply:
Done.

Reply:
Done.

> Couple comments/questions, I like the direction and am happy to see the improvements.
> 
> Side-note for out of band discussion, I think when we introduced PerfTimer it was a more or less explicit decision the timing code would not be left in. I am not sure we do want ever want to disable and not remove timing code. Timing code usually tends to be pretty one-off, what's more if you enabled it and hten suddenly there were a few dozen timers, it would create a lot of noise in the logs. I would also wonder if at some point we might see accounting/memory issues if we are simply suppressing output.

My thinking is it is useful to instrument some places in the code and check it in, but with a 0 reporting frequency by default for these reasons:

- If a developer wants to reproduce a past result, the code is already there. They don't have to insert the code themselves, potentially in a slightly different way than before.
- If we make some prefs around enabling some existing timers, it can allow users to report their results with a release version without code changes. For example, if AI is slow on a specific map, they can enable the timing category for that (which will make the code pass a non-0 reporting frequency to the perftimers) and report their output.
- It allows re-use of effort when finding the right place to instrument code. For example, instrumenting the AI code in the right way is non-trivial - e.g. you want to measure the time the AI takes thinking, but not the time used when there's a battle popup and a human player needs to make a decision during an AI's combat phase. If we check in these timers, finding the right place to instrument these things can be done once, instead of many times.

By minimizing the overhead when frequency is set to 0 (i.e. not querying the current time or doing any computation), we can have perf timers added to interesting places in the code providing the above benefits.

Reply:
Interesting points. I think the counter is how frequent we would actually re-enable timers compared to the "clutter" it creates. I think if we want sustained measurements it would be better to bring in perhaps AOP and some annotations to do the timing. The timers affect variable scope, which makes for non-zero clutter.

I'm still concerned as well if you turn on timers then you have many timers start firing and it can be hard to see the output you want.

I think any timing for confirmation testing probably should live in actual test code and be a bona fide performance test. I've always chafed that we sometimes reject PRs or make PRs suddenly have to do timing measurements when any such requirements are ideally baked into the test suite.

Perhaps we can make some things easier to encapsulating specific sections of the AI code in a way that makes it easy to time. Then if we document those locations nicely we could get a few benefits beyond easy timing.